### PR TITLE
Add list-monitors command.

### PR DESCRIPTION
**Pull request by glogiotatidis**

### CODE REVIEW

**Comment by glogiotatidis** [comment]
OK this is turning into gigantic pull request as the core of the tool get build. There's still work to add tests (by tomorrow ™️) and fix the packaging which causes the failing Travis job.

Other than that you can go crazy with code review @akatsoulas @panospet @nmakro 

Thanks!

**Comment by akatsoulas** [comment]
@glogiotatidis can you add a couple of comments about this PR and maybe some links to documentation to get a better understanding on what you are building?

**Comment by glogiotatidis** [comment]
> @glogiotatidis can you add a couple of comments about this PR and maybe some links to documentation to get a better understanding on what you are building?

it's super obvious ( ❗️ ) but yes. tomorrow.

**Comment by glogiotatidis** [comment]
OK @akatsoulas here you go, [some basic docs to get you started](https://github.com/glogiotatidis/neres/blob/f3b548619426bda84ab96dc017dd64a01d94132f/README.md). Please pinpoint any pain problems during install / use to fix them. Thanks!

File: neres/cli.py [hunk 1]
```diff
@@ -1,15 +1,313 @@
...
+@click.option('--frequency', default=None)
+@click.option('--sla-threshold', 'slaThreshold', default=None)
+@click.option('--validation-string', default=None)
+@click.option('--no-validation-string', default=None, is_flag=True)
+@click.option('--bypass-head-request/--no-bypass-head-request', default=None, is_flag=True)
+@click.option('--verify-ssl/--no-verify-ssl', default=None, is_flag=True)
+@click.option('--redirect-is-failure/--no-redirect-is-failure', default=None, is_flag=True)
+@click.option('--raw', default=False, is_flag=True)
+def update_monitor(monitor, **kwargs):
+    if kwargs['no_validation_string']:
```
Comment (akatsoulas) [inline neres/cli.py]: nit: I would go with

```python
no_validation_string = kwargs.get('no_validation_string')
if no_validation_string:
    do_stuff
```
This way you can also make sure that you won't get a `KeyError` exception.

File: neres/newrelic.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+
+def _construct_metadata(validation_string=None, bypass_head_request=None,
+                        verify_ssl=None, redirect_is_failure=None,
+                        metadata=None):
+    if metadata is None:
+        metadata = {}
+
+    if bypass_head_request is True:
+        metadata['nr.synthetics.metadata.job.options.simple.bypass.head'] = True
+    elif bypass_head_request is False:
```
Comment (akatsoulas) [inline neres/newrelic.py:162]: You could just replace this with an `else` clause.

File: neres/session.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+            self.xsrf_token = xsrf_token
+
+    def _set_xsrf_headers(self, kwargs):
+        xsrf_header = {
+            'X-XSRF-TOKEN': self.xsrf_token,
+        }
+        if 'headers' in kwargs:
+            kwargs['headers'].update(xsrf_header)
+        else:
+            kwargs['headers'] = xsrf_header
```
Comment (akatsoulas) [inline neres/session.py:18]: Probably the `if/else` could be written as kwargs['headers'] = xsrf_header

File: setup.cfg [hunk 4]
```diff
@@ -16,3 +16,4 @@ universal = 1
 
 [flake8]
 exclude = docs
+max-line-length=100
```
Comment (akatsoulas) [inline setup.cfg]: s/100/99

**Review by akatsoulas — COMMENTED** [review]
I did a quick pass and added a few nit comments. I will set this up locally and have a more thorough look today/tomorrow.

File: neres/cli.py (hunk 1 above)
Comment (glogiotatidis) [reply neres/cli.py]: `click` will always populate the variables with default or user provided values, so the key will by always there.

File: neres/newrelic.py (hunk 2 above)
Comment (glogiotatidis) [reply neres/newrelic.py:162]: bypass_head_request can be `None`, in the case `update` where I don't want to touch it

File: neres/session.py (hunk 3 above)
Comment (glogiotatidis) [reply neres/session.py:18]: If headers are already present I want to amend the xsrf token and not replace everything

File: setup.cfg (hunk 4 above)
Comment (glogiotatidis) [reply setup.cfg]: ok ok

**Comment by glogiotatidis** [comment]
hey folks, I added a login command and refactored a couple of things. Please get in your comments by EOD if possible.

thanks!

**Comment by glogiotatidis** [comment]
BTW the tests are pretty default from the cookiecutter template right now, I wouldn't spend a second reviewing them.

**Comment by panospet** [comment]
I don't have many comments about the code itself, as I don't have much familiarity with the things you do. One notice and one bug I discovered so far, though: 

1. Open command needs the id as argument. I would prefer the monitor name to run it, instead of doing list-monitors first, and then copy the id to reuse it.
2. I think something broke in your last commit, the one you tried to enhance authorization. See:
```
(neres) panos@pp:~/neres$ neres list-monitors
┌Monitors───┬───────────────┬──────────────────────────────────────┬─────────┬───────────┬───────────┬───────────┬───────────┬───────┬────────┐
│ # │ H │ S │ Name          │ ID                                   │ Success │  Avg Size │ Load time │ Load time │ Frequency │  Loca │  Notif │
│   │   │   │               │                                      │    Rate │           │ (50th PR) │ (95th PR) │           │ tions │ Emails │
├───┼───┼───┼───────────────┼──────────────────────────────────────┼─────────┼───────────┼───────────┼───────────┼───────────┼───────┼────────┤
│ 1 │ ✔ │ ✔ │ panos-monitor │ 93fa4edb-5a3d-4bbe-ac6f-41f31fdc1d8f │  100.0% │ 770 Bytes │ 1159.0 ms │ 1341.9 ms │    10 min │     1 │      1 │
└───┴───┴───┴───────────────┴──────────────────────────────────────┴─────────┴───────────┴───────────┴───────────┴───────────┴───────┴────────┘
(neres) panos@pp:~/neres$ neres open 93fa4edb-5a3d-4bbe-ac6f-41f31fdc1d8f
Traceback (most recent call last):
  File "/home/panos/Envs/neres/bin/neres", line 11, in <module>
    load_entry_point('neres==0.1.0', 'console_scripts', 'neres')()
  File "/home/panos/Envs/neres/local/lib/python2.7/site-packages/click-6.7-py2.7.egg/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/home/panos/Envs/neres/local/lib/python2.7/site-packages/click-6.7-py2.7.egg/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/home/panos/Envs/neres/local/lib/python2.7/site-packages/click-6.7-py2.7.egg/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/panos/Envs/neres/local/lib/python2.7/site-packages/click-6.7-py2.7.egg/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/panos/Envs/neres/local/lib/python2.7/site-packages/click-6.7-py2.7.egg/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/home/panos/Envs/neres/local/lib/python2.7/site-packages/neres-0.1.0-py2.7.egg/neres/cli.py", line 294, in open_monitor
    url = urls.MONITOR.format(monitor)
KeyError: 'account'

```

**Comment by glogiotatidis** [comment]
Thanks for your time @akatsoulas @panospet 🍻

**Comment by panospet** [comment]
Always at your service :)

**Pull request merged**